projectroles/forms.py

```diff
@@ -541,9 +553,86 @@ def __init__(self, project=None, current_user=None, *args, **kwargs):
             self.fields['description'].widget = forms.HiddenInput()
             self.fields['readme'].widget = forms.HiddenInput()
 
+    @classmethod
+    def _validate_app_settings(
+        self,
+        cleaned_data,
+        app_plugins,
+        app_settings,
+        p_kwargs,
+        instance,
+        instance_owner_as,
+    ):
+        """Validate and clean app_settings form fields"""
+        errors = []
+
+        if cleaned_data.get('type') == PROJECT_TYPE_CATEGORY:
+            return (cleaned_data, errors)
+
+        for plugin in app_plugins + [None]:
+            if plugin:
+                name = plugin.name
+                p_settings = app_settings.get_definitions(
+                    APP_SETTING_SCOPE_PROJECT, app_name=name, **p_kwargs
+                )
+            else:
+                name = 'projectroles'
+                p_settings = app_settings.get_definitions(
+                    APP_SETTING_SCOPE_PROJECT, app_name=name, **p_kwargs
+                )
+
+            plugin_app_settings = {}
+            for s_key, s_val in p_settings.items():
+                s_field = 'settings.{}.{}'.format(name, s_key)
+                plugin_app_settings[s_key] = cleaned_data.get(s_field)
+
+                if s_val['type'] == 'JSON':
+                    if s_val['type'] == 'JSON':
+                        if not cleaned_data.get(s_field):
+                            cleaned_data[s_field] = '{}'
+                        try:
+                            cleaned_data[s_field] = json.loads(
+                                cleaned_data.get(s_field)
+                            )
+                        except json.JSONDecodeError as err:
+                            errors.append(
+                                (s_field, 'Invalid JSON\n' + str(err))
+                            )
+                elif s_val['type'] == 'INTEGER':
+                    # When the field is a select/dropdown the information of
+                    # the datatype gets lost. We need to convert that here,
+                    # otherwise subsequent checks will fail.
+                    cleaned_data[s_field] = int(cleaned_data[s_field])
+
+                if not app_settings.validate(
+                    setting_type=s_val['type'],
+                    setting_value=cleaned_data.get(s_field),
+                    setting_options=s_val.get('options'),
+                ):
+                    errors.append((s_field, 'Invalid value'))
+
+            # Custom validation for app settings
+            try:
+                app_settings_errors = plugin.validate_app_settings(
+                    plugin_app_settings,
+                    project=instance,
+                    user=instance_owner_as,
+                )
```

Comment:
Here I pass in `plugin.validate_app_settings` such data:
for `bgjobs` - {}
for `example_project_app`  - {'project_str_setting': '', 'project_int_setting': 0, 'project_str_setting_options': 'string1', 'project_int_setting_options': '0', 'project_bool_setting': True, 'project_global_setting': False, 'project_json_setting': '{"list": [1, 2, 3, 4, 5], "Example": "Value", "level_6": false}', 'project_hidden_setting': '', 'project_hidden_json_setting': '{}'}
for `filesfolders` - {'allow_public_links': False}
for `timeline` - {}
and for `projectroles` - {'ip_restrict': False, 'ip_allowlist': '{}'}

Is it what supposed to be?

Reply:
That looks exactly correct :)